### PR TITLE
chore: optimizing ordered loop

### PR DESCRIPTION
**Pull request by ashwinidulams**
optimization to the ordered loop so that queueing and dequeuing continues to stay at O(1). 

we can delay the enqueuing of the task until the close of the book as the desired order is guaranteed.

### CODE REVIEW

File: pkg/reduce/readloop/readloop.go [hunk 1]
```diff
@@ -362,6 +366,9 @@ func (rl *ReadLoop) closePartitions(partitions []partition.ID) {
 	for _, p := range partitions {
 		q := rl.pbqManager.GetPBQ(p)
 		rl.log.Infow("Close of book", zap.String("partitionID", p.String()))
+		rl.op.insertTask(rl.udfInvocationTracking[p])
+		partitionsInFlight.With(map[string]string{metrics.LabelVertex: rl.op.vertexName, metrics.LabelPipeline: rl.op.pipelineName}).Inc()
 		q.CloseOfBook()
```
Comment (yhl25) [inline pkg/reduce/readloop/readloop.go:371]: why are we incrementing partitionsInFlight while closing the partition? merge conflict?

Comment (ashwinidulams) [reply pkg/reduce/readloop/readloop.go:371]: i moved it here, but i guess i can leave it there itself.

**Review by yhl25 — APPROVED** [review]

File: pkg/reduce/readloop/readloop.go (hunk 1 above)
Comment (ashwinidulams) [reply pkg/reduce/readloop/readloop.go]: since this is streaming, it makes sense to increment this as soon as the go routine is scheduled.

Comment (yhl25) [reply pkg/reduce/readloop/readloop.go]: yeah

**Review by vigith — APPROVED** [review]

**Pull request merged**